File: src/test/java/kitchenpos/products/tobe/domain/ProductTest.java
package kitchenpos.products.tobe.domain;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import java.math.BigDecimal;

import static org.assertj.core.api.Assertions.assertThat;

class ProductTest {

    @DisplayName("상품 가격을 변경한다.")
    @ParameterizedTest
    @ValueSource(strings = {"0", "1", "1000"})
    void changePrice(BigDecimal price) {
        // given
        var product = new Product();
        var productPrice = new ProductPrice(price);

        // when
        product.changePrice(productPrice);

        // then
        assertThat(product.getPrice()).isEqualTo(productPrice);
    }
}

File: src/test/java/kitchenpos/products/tobe/domain/ProductTest.kt
package kitchenpos.products.tobe.domain

import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.types.shouldBeInstanceOf

class ProductTest : BehaviorSpec({
    Given("상품 가격을 변경할 떄") {
        val name = ProductName("테스트 상품") { true }
        val price = ProductPrice.valueOf(1000)
        val product = Product(name, price)

        When("상품 가격을 변경하면") {
            val newPrice = ProductPrice.valueOf(2000)
            product.changePrice(newPrice)

            Then("상품 가격이 변경된다") {
                product.price shouldBe newPrice
            }

            Then("도메인 이벤트가 발생한다") {
                val domainEvents = product.getAndClearDomainEvents()

                domainEvents.size shouldBe 1
                domainEvents[0].shouldBeInstanceOf<ProductPriceChangedEvent>()
            }
        }
    }
})
